transcom/mymove
# B 18003 okta looping error fix

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A869812&RoomContext=TeamRoom%3A808731&concept=TeamRoom)

## Summary

Office & Admin users that have been authenticating with Okta into MilMove have been reporting issues with Okta looping them back to the home page.

This is likely due to an active session already existing in the browser and the user was likely not signing out of MilMove which fires logging them out of Okta so although their MilMove browser was closed, their Okta session continues to be active. If the user did not sign out of MM or Okta and they open a new tab and try to log into MilMove again... This will loop them back to the redirect URI since they used their only authenticator (CAC) and they can't use that same authenticator again since they have an active session.

When the user does not have any more authenticators to use when authenticating, Okta sends back an `error_description` in the url with the value `"The resource owner or authorization server denied the request."`. This BL item focuses on detecting that error message and handling the logic for two different scenarios:

- If the user closed the MM tab and didn't log out, they should still have their okta `ID_Token` saved to `appCtx`, which we can use to log them out of Okta, redirect them back to the home page with an info banner, and have them log in again successfully
- If the user logged into Okta elsewhere **prior** to logging into MM, they have an active Okta session, but they don't have their ID token saved to `appCtx`, so we will be unable to log them out since we can't identify them from MilMove during the authentication handshake. What we will do though, is conditional logic that redirects the user back to the home page with params `okta_logged_out` being either `true` or `false` and displays respective info banners for each.

Some small UI work was done in this story that focused on creating an Okta specific banner that shows when they are redirected following logout when MM attaches the param `okta_logged_out=true` or `okta_logged_out=false` to the redirect url.

If you want to just see the banners, you can input the params into the URL:
`http://officelocal:3000/sign-in?okta_logged_out=true`
or
`http://officelocal:3000/sign-in?okta_logged_out=false`

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Access the MM app as an office/admin user
2. Login and authenticate with Okta
3. Close your tab or whatever to end the MM session (can also clear the session cookie)
4. Sign into MM again
5. You'll see the info banner


## Screenshots
When the user has the ID token still saved to `appCtx` and is asked to authenticate with Okta, we can log them out.
![Screenshot 2023-12-19 at 12 12 39 PM](https://github.com/transcom/mymove/assets/136510600/6729f241-283a-4a1c-a3bc-3cd378969edc)

If we don't have the ID token still, we will need the user to log out of Okta themselves.
![Screenshot 2023-12-19 at 12 12 30 PM](https://github.com/transcom/mymove/assets/136510600/aeef4beb-434b-4f56-ad79-1b975cc4cc05)
